src/toil/wdl/wdl_functions.py

```diff
@@ -957,3 +957,22 @@ def keys(in_map: dict) -> list:
     """
 
     return list(in_map.keys())
+
+
+def collect_by_key(in_array: List[WDLPair]) -> dict:
+    """
+    Given an Array consisting of Pairs, the `collect_by_key` function returns a Map
+    in which the left elements of the Pairs are the keys and the right elements the
+    values.
+
+    WDL syntax: Map[X,Array[Y]] collect_by_key(Array[Pair[X,Y]])
+    """
+    if not isinstance(in_array, list):
+        raise WDLRuntimeError(f'as_map() requires "{in_array}" to be a list!  Not: {type(in_array)}')
+
+    map = {}
+
+    for pair in in_array:
+        map.setdefault(pair.left, []).append(pair.right)
```

Comment:
Oh, I haven't seen this before.  This is nice.  The alternative that I usually see (just to make you aware of it, no need to change this), is `collections.defaultdict(list)`: https://docs.python.org/3/library/collections.html#defaultdict-examples